File: tests_ext/src/test/java/com/dirgub/klee0kai/stone/text_ext/cache/EarthCacheProMixedTests.java
package com.dirgub.klee0kai.stone.text_ext.cache;

import com.github.klee0kai.stone.Stone;
import com.github.klee0kai.test.di.gcforest.GcGodComponent;
import com.github.klee0kai.test.mowgli.earth.Mountain;
import com.github.klee0kai.test_ext.inject.di.gcforest.GcGodExtComponent;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotEquals;

public class EarthCacheProMixedTests {


    @Test
    public void strongCacheTest() {
        //Given
        GcGodComponent DI = Stone.createComponent(GcGodComponent.class);
        GcGodExtComponent DIPro = Stone.createComponent(GcGodExtComponent.class);
        DIPro.extOf(DI);

        //When
        Mountain mountain1 = DIPro.earth().mountainStrong();
        Mountain mountain2 = DI.earth().mountainStrong();

        //Then
        assertEquals(
                mountain1.uuid,
                mountain2.uuid
        );
    }


    @Test
    public void softCacheTest() {
        //Given
        GcGodComponent DI = Stone.createComponent(GcGodComponent.class);
        GcGodExtComponent DIPro = Stone.createComponent(GcGodExtComponent.class);
        DIPro.extOf(DI);

        //When
        Mountain mountain1 = DIPro.earth().mountainSoft();
        Mountain mountain2 = DI.earth().mountainSoft();

        //Then
        assertEquals(
                mountain1.uuid,
                mountain2.uuid
        );
    }

    @Test
    public void weakCacheTest() {
        //Given
        GcGodComponent DI = Stone.createComponent(GcGodComponent.class);
        GcGodExtComponent DIPro = Stone.createComponent(GcGodExtComponent.class);
        DIPro.extOf(DI);

        //When
        Mountain mountain1 = DIPro.earth().mountainWeak();
        Mountain mountain2 = DI.earth().mountainWeak();

        //Then
        assertEquals(
                mountain1.uuid,
                mountain2.uuid
        );
    }

    @Test
    public void defCacheTest() {
        //Given
        GcGodComponent DI = Stone.createComponent(GcGodComponent.class);
        GcGodExtComponent DIPro = Stone.createComponent(GcGodExtComponent.class);
        DIPro.extOf(DI);

        //When
        Mountain mountain1 = DIPro.earth().mountainDefaultFactory();
        Mountain mountain2 = DI.earth().mountainDefaultFactory();

        //Then
        assertNotEquals(
                mountain1.uuid,
                mountain2.uuid,
                "Factory providing"
        );
    }

    @Test
    public void defCache2Test() {
        //Given
        GcGodComponent DI = Stone.createComponent(GcGodComponent.class);
        GcGodExtComponent DIPro = Stone.createComponent(GcGodExtComponent.class);
        DIPro.extOf(DI);

        //When
        Mountain mountain1 = DIPro.earth().mountainDefault2Factory();
        Mountain mountain2 = DI.earth().mountainDefault2Factory();

        //Then
        assertNotEquals(
                mountain1.uuid,
                mountain2.uuid,
                "Factory providing"
        );
    }


    @Test
    public void differentMethodDifferentCacheTest() {
        //Given
        GcGodComponent DI = Stone.createComponent(GcGodComponent.class);
        GcGodExtComponent DIPro = Stone.createComponent(GcGodExtComponent.class);
        DIPro.extOf(DI);

        //When
        Mountain mountainStrong = DIPro.earth().mountainStrong();
        Mountain mountainSoft = DI.earth().mountainSoft();

        //Then
        assertNotEquals(
                mountainStrong.uuid,
                mountainSoft.uuid
        );
    }


    @Test
    public void factoryNotCacheTest() {
        //Given
        GcGodComponent DI = Stone.createComponent(GcGodComponent.class);
        GcGodExtComponent DIPro = Stone.createComponent(GcGodExtComponent.class);
        DIPro.extOf(DI);

        //When
        Mountain mountain1 = DIPro.earth().mountainFactory();
        Mountain mountain2 = DI.earth().mountainFactory();

        //Then
        assertNotEquals(
                mountain1.uuid,
                mountain2.uuid
        );
    }


    @Test
    public void differentDIDifferentCacheTest() {
        //Given
        GcGodComponent DI1 = Stone.createComponent(GcGodComponent.class);
        GcGodExtComponent DIPro1 = Stone.createComponent(GcGodExtComponent.class);
        DIPro1.extOf(DI1);
        GcGodComponent DI2 = Stone.createComponent(GcGodComponent.class);
        GcGodExtComponent DIPro2 = Stone.createComponent(GcGodExtComponent.class);
        DIPro2.extOf(DI2);

        //When
        Mountain mountain1 = DIPro1.earth().mountainStrong();
        Mountain mountain2 = DI2.earth().mountainStrong();

        //Then
        assertNotEquals(
                mountain1.uuid,
                mountain2.uuid
        );
    }

}

File: tests_ext/src/test/java/com/dirgub/klee0kai/stone/text_ext/cache/EarthCacheProMixedTests.kt
package com.dirgub.klee0kai.stone.text_ext.cache

import com.github.klee0kai.stone.Stone
import com.github.klee0kai.test.di.gcforest.GcGodComponent
import com.github.klee0kai.test.di.gcforest.GcGodComponentStoneComponent
import com.github.klee0kai.test.mowgli.earth.Mountain
import com.github.klee0kai.test_ext.inject.di.gcforest.GcGodExtComponent
import com.github.klee0kai.test_ext.inject.di.gcforest.GcGodExtComponentStoneComponent
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertNotEquals
import org.junit.jupiter.api.Test

class EarthCacheProMixedTests {

    @Test
    fun strongCacheTest() {
        //Given
        val DI: GcGodComponent = GcGodComponentStoneComponent()
        val DIPro: GcGodExtComponent = GcGodExtComponentStoneComponent()
        DIPro.extOf(DI)

        //When
        val mountain1: Mountain? = DIPro.earth()!!.mountainStrong()
        val mountain2: Mountain? = DI.earth()!!.mountainStrong()

        //Then
        assertEquals(
            mountain1!!.uuid,
            mountain2!!.uuid
        )
    }


    @Test
    fun softCacheTest() {
        //Given
        val DI: GcGodComponent = GcGodComponentStoneComponent()
        val DIPro: GcGodExtComponent = GcGodExtComponentStoneComponent()
        DIPro.extOf(DI)

        //When
        val mountain1: Mountain? = DIPro.earth()!!.mountainSoft()
        val mountain2: Mountain? = DI.earth()!!.mountainSoft()

        //Then
        assertEquals(
            mountain1!!.uuid,
            mountain2!!.uuid
        )
    }

    @Test
    fun weakCacheTest() {
        //Given
        val DI: GcGodComponent = GcGodComponentStoneComponent()
        val DIPro: GcGodExtComponent = GcGodExtComponentStoneComponent()
        DIPro.extOf(DI)

        //When
        val mountain1: Mountain? = DIPro.earth()!!.mountainWeak()
        val mountain2: Mountain? = DI.earth()!!.mountainWeak()

        //Then
        assertEquals(
            mountain1!!.uuid,
            mountain2!!.uuid
        )
    }

    @Test
    fun defCacheTest() {
        //Given
        val DI: GcGodComponent = GcGodComponentStoneComponent()
        val DIPro: GcGodExtComponent = GcGodExtComponentStoneComponent()
        DIPro.extOf(DI)

        //When
        val mountain1: Mountain? = DIPro.earth()!!.mountainDefaultFactory()
        val mountain2: Mountain? = DI.earth()!!.mountainDefaultFactory()

        //Then
        assertNotEquals(
            mountain1!!.uuid,
            mountain2!!.uuid,
            "Factory providing"
        )
    }

    @Test
    fun defCache2Test() {
        //Given
        val DI: GcGodComponent = GcGodComponentStoneComponent()
        val DIPro: GcGodExtComponent = GcGodExtComponentStoneComponent()
        DIPro.extOf(DI)

        //When
        val mountain1: Mountain? = DIPro.earth()!!.mountainDefault2Factory()
        val mountain2: Mountain? = DI.earth()!!.mountainDefault2Factory()

        //Then
        assertNotEquals(
            mountain1!!.uuid,
            mountain2!!.uuid,
            "Factory providing"
        )
    }


    @Test
    fun differentMethodDifferentCacheTest() {
        //Given
        val DI: GcGodComponent = GcGodComponentStoneComponent()
        val DIPro: GcGodExtComponent = GcGodExtComponentStoneComponent()
        DIPro.extOf(DI)

        //When
        val mountainStrong: Mountain? = DIPro.earth()!!.mountainStrong()
        val mountainSoft: Mountain? = DI.earth()!!.mountainSoft()

        //Then
        assertNotEquals(
            mountainStrong!!.uuid,
            mountainSoft!!.uuid
        )
    }


    @Test
    fun factoryNotCacheTest() {
        //Given
        val DI: GcGodComponent = GcGodComponentStoneComponent()
        val DIPro: GcGodExtComponent = GcGodExtComponentStoneComponent()
        DIPro.extOf(DI)

        //When
        val mountain1: Mountain? = DIPro.earth()!!.mountainFactory()
        val mountain2: Mountain? = DI.earth()!!.mountainFactory()

        //Then
        assertNotEquals(
            mountain1!!.uuid,
            mountain2!!.uuid
        )
    }


    @Test
    fun differentDIDifferentCacheTest() {
        //Given
        val DI1: GcGodComponent = GcGodComponentStoneComponent()
        val DIPro1: GcGodExtComponent = GcGodExtComponentStoneComponent()
        DIPro1.extOf(DI1)
        val DI2: GcGodComponent = GcGodComponentStoneComponent()
        val DIPro2: GcGodExtComponent = GcGodExtComponentStoneComponent()
        DIPro2.extOf(DI2)

        //When
        val mountain1: Mountain? = DIPro1.earth()!!.mountainStrong()
        val mountain2: Mountain? = DI2.earth()!!.mountainStrong()

        //Then
        assertNotEquals(
            mountain1!!.uuid,
            mountain2!!.uuid
        )
    }

}
